hibernate-core/src/main/java/org/hibernate/engine/internal/Cascade.java

```diff
@@ -87,21 +88,20 @@ public static void cascade(
 			final boolean hasUninitializedLazyProperties = persister.hasUninitializedLazyProperties( parent );
 			final int componentPathStackDepth = 0;
 			for ( int i = 0; i < types.length; i++) {
-				final CascadeStyle style = cascadeStyles[i];
-				final String propertyName = propertyNames[i];
-
-				if ( style.doCascade( action ) ) {
-					Object child;
+				final CascadeStyle style = cascadeStyles[ i ];
+				final String propertyName = propertyNames[ i ];
 
+				Object child = null;
+				if ( style.doCascade( action ) || action.deleteOrphans() ) {
 					if ( hasUninitializedLazyProperties &&
 							!persister.getInstrumentationMetadata().isAttributeLoaded( parent, propertyName ) ) {
 						// parent is a bytecode enhanced entity.
 						// cascading to an uninitialized, lazy value.
-						if ( types[i].isCollectionType() ) {
+						if ( types[ i ].isCollectionType() ) {
 							// The collection does not need to be loaded from the DB.
 							// CollectionType#resolve will return an uninitialized PersistentCollection.
 							// The action will initialize the collection later, if necessary.
-							child = types[i].resolve( LazyPropertyInitializer.UNFETCHED_PROPERTY, eventSource, parent );
+							child = types[ i ].resolve( LazyPropertyInitializer.UNFETCHED_PROPERTY, eventSource, parent );
 							// TODO: it would be nice to be able to set the attribute in parent using
 							// persister.setPropertyValue( parent, i, child ).
 							// Unfortunately, that would cause the uninitialized collection to be
```

Comment:
When changing the if-structure to:
<pre>
if ( style.doCascade( action ) ) {
  cascadeProperty( ... );
} 
else if ( !style.doCascade( action ) && action.deleteOrphans() ) {
  cascadeLogicalOneToOneOrphanRemoval( ... );
}
else if ( action.requiresNoCascadeChecking() ) {
  action.noCascade( ... );
}
</pre>
This won't work for `org.hibernate.jpa.test.orphan.onetoone.OneToOneOrphanTest` where you have a transient value assigned to a property.

As mentioned below, `PERSIST_ON_FLUSH`'s cascade action is designed to return false on the first if-check, but the problem is that the first else-if check is valid for this action too and therefore the expected invocation of `action.noCascade(...)` never happens.  

That is why the code was structured as:
<pre>
if ( style.doCascade( action ) ) {
  cascadeProperty( ... );
}
else if ( action.requiresNoCascadeChecking() ) {
  action.noCascade( ... );
  cascadeLogicalOneToOneOrphanRemoval( ... );
}
</pre>

For the situation of `OneToOneOrphanTest`, the `action.noCascade` call would assert throwing an exception, keeping the intended behavior in tact.  But for situations where the properties weren't transient and thus no exception thrown, the one-to-one orphan removal would fire.

As for the lazy initialization, we require that that `Child` be initialized under two scenarios:
1. style.doCascade( action ) == true
2. action.deleteOrphans() == true

To satisfy those two requirements, I simply moved the initialization block for the Child object out of the original inner if so that it could be used for both scenarios without having to duplicate the code.

Perhaps I am missing something?